### PR TITLE
feat(manager/bundler): support groups within source blocks

### PR DESCRIPTION
**Pull request by mustardnoise**
## Changes

This PR fixes the updating of gems from private sources with groups.

## Context

I have a Gemfile with the private source and groups within this source.

```ruby
source 'https://rubygems.org'

ruby '3.3.0'

gem 'rails', '~> 7.1'

source 'https://PRIVATE' do
  group :test do
    gem 'private_dummy', '~> 0.0.1'
  end

  group :production do
    gem 'another_private_dummy', '~> 0.0.1'
  end
end
```

It is currently not possible to update gems in the next group(s) within the private source.

Example of a log message **BEFORE** the fix: 

```
DEBUG: packageFiles with updates (repository=playground/renovate_bug, baseBranch=main)
       "config": {
         "bundler": [
           {
             "registryUrls": ["https://rubygems.org"],
             "deps": [
               {
                 "depName": "ruby",
                 "currentValue": "3.3.0",
                 "datasource": "ruby-version",
                 "registryUrls": null,
                 "updates": [],
                 "packageName": "ruby",
                 "versioning": "ruby",
                 "warnings": [],
                 "sourceUrl": "https://github.com/ruby/ruby",
                 "registryUrl": "https://www.ruby-lang.org/",
                 "homepage": "https://www.ruby-lang.org",
                 "currentVersion": "3.3.0",
                 "fixedVersion": "3.3.0"
               },
               {
                 "depName": "rails",
                 "managerData": {"lineNumber": 4},
                 "currentValue": "'~> 7.1'",
                 "datasource": "rubygems",
                 "lockedVersion": "7.1.3",
                 "updates": [],
                 "packageName": "rails",
                 "versioning": "ruby",
                 "warnings": [],
                 "sourceUrl": "https://github.com/rails/rails",
                 "registryUrl": "https://rubygems.org",
                 "homepage": "https://rubyonrails.org",
                 "changelogUrl": "https://github.com/rails/rails/releases/tag/v7.1.3",
                 "currentVersion": "7.1.3",
                 "fixedVersion": "7.1.3"
               },
               {
                 "depName": "private_dummy",
                 "managerData": {"lineNumber": 8},
                 "currentValue": "'~> 0.0.1'",
                 "datasource": "rubygems",
                 "depTypes": ["test"],
                 "registryUrls": [
                   "https://PRIVATE"
                 ],
                 "lockedVersion": "0.0.1",
                 "updates": [
                   {
                     "bucket": "non-major",
                     "newVersion": "0.0.2",
                     "newValue": "'~> 0.0.1'",
                     "releaseTimestamp": "2024-02-19T15:55:40.834Z",
                     "newMajor": 0,
                     "newMinor": 0,
                     "updateType": "patch",
                     "isRange": true,
                     "isLockfileUpdate": true,
                     "branchName": "renovate/all-minor-patch"
                   }
                 ],
                 "packageName": "private_dummy",
                 "versioning": "ruby",
                 "warnings": [],
                 "sourceUrl": "https://PRIVATE/private_dummy",
                 "registryUrl": "https://PRIVATE",
                 "currentVersion": "0.0.1",
                 "isSingleVersion": true,
                 "fixedVersion": "0.0.1"
               },
               {
                 "depName": "another_private_dummy",
                 "managerData": {"lineNumber": 12},
                 "currentValue": "'~> 0.0.1'",
                 "datasource": "rubygems",
                 "lockedVersion": "0.0.1",
                 "updates": [],
                 "packageName": "another_private_dummy",
                 "versioning": "ruby",
                 "warnings": [
                   {
                     "topic": "another_private_dummy",
                     "message": "Failed to look up rubygems package another_private_dummy"
                   }
                 ]
               }
             ],
             "lockFiles": ["Gemfile.lock"],
             "packageFile": "Gemfile"
           }
         ]
       }

...

 WARN: Package lookup failures (repository=playground/renovate_bug, baseBranch=main, branch=renovate/all-minor-patch)
       "warnings": ["Failed to look up rubygems package another_private_dummy"],
       "files": ["Gemfile"]

...

DEBUG: branches info extended (repository=playground/renovate_bug)
       "branchesInformation": [
         {
           "branchName": "renovate/all-minor-patch",
           "prNo": 2,
           "prTitle": "feat(renovate): Update dependency private_dummy to v0.0.2",
           "result": "done",
           "upgrades": [
             {
               "datasource": "rubygems",
               "depName": "private_dummy",
               "displayPending": "",
               "fixedVersion": "0.0.1",
               "currentVersion": "0.0.1",
               "currentValue": "'~> 0.0.1'",
               "newValue": "'~> 0.0.1'",
               "newVersion": "0.0.2",
               "packageFile": "Gemfile",
               "updateType": "patch",
               "packageName": "private_dummy"
             }
           ]
         }
       ]
```

Example of a log message **AFTER** the fix: 

```
DEBUG: packageFiles with updates (repository=playground/renovate_bug, baseBranch=main)
       "config": {
         "bundler": [
           {
             "registryUrls": ["https://rubygems.org"],
             "deps": [
               {
                 "depName": "ruby",
                 "currentValue": "3.3.0",
                 "datasource": "ruby-version",
                 "registryUrls": null,
                 "updates": [],
                 "packageName": "ruby",
                 "versioning": "ruby",
                 "warnings": [],
                 "sourceUrl": "https://github.com/ruby/ruby",
                 "registryUrl": "https://www.ruby-lang.org/",
                 "homepage": "https://www.ruby-lang.org",
                 "currentVersion": "3.3.0",
                 "fixedVersion": "3.3.0"
               },
               {
                 "depName": "rails",
                 "managerData": {"lineNumber": 4},
                 "currentValue": "'~> 7.1'",
                 "datasource": "rubygems",
                 "lockedVersion": "7.1.3",
                 "updates": [],
                 "packageName": "rails",
                 "versioning": "ruby",
                 "warnings": [],
                 "sourceUrl": "https://github.com/rails/rails",
                 "registryUrl": "https://rubygems.org",
                 "homepage": "https://rubyonrails.org",
                 "changelogUrl": "https://github.com/rails/rails/releases/tag/v7.1.3",
                 "currentVersion": "7.1.3",
                 "fixedVersion": "7.1.3"
               },
               {
                 "depName": "private_dummy",
                 "managerData": {"lineNumber": 8},
                 "currentValue": "'~> 0.0.1'",
                 "datasource": "rubygems",
                 "depTypes": ["test"],
                 "registryUrls": [
                   "https://PRIVATE"
                 ],
                 "lockedVersion": "0.0.1",
                 "updates": [
                   {
                     "bucket": "non-major",
                     "newVersion": "0.0.2",
                     "newValue": "'~> 0.0.1'",
                     "releaseTimestamp": "2024-02-19T15:55:40.834Z",
                     "newMajor": 0,
                     "newMinor": 0,
                     "updateType": "patch",
                     "isRange": true,
                     "isLockfileUpdate": true,
                     "branchName": "renovate/all-minor-patch"
                   }
                 ],
                 "packageName": "private_dummy",
                 "versioning": "ruby",
                 "warnings": [],
                 "sourceUrl": "https://PRIVATE/private_dummy",
                 "registryUrl": "https://PRIVATE",
                 "currentVersion": "0.0.1",
                 "isSingleVersion": true,
                 "fixedVersion": "0.0.1"
               },
               {
                 "depName": "another_private_dummy",
                 "managerData": {"lineNumber": 12},
                 "currentValue": "'~> 0.0.1'",
                 "datasource": "rubygems",
                 "depTypes": ["production"],
                 "registryUrls": [
                   "https://PRIVATE"
                 ],
                 "lockedVersion": "0.0.1",
                 "updates": [
                   {
                     "bucket": "non-major",
                     "newVersion": "0.0.2",
                     "newValue": "'~> 0.0.1'",
                     "releaseTimestamp": "2024-02-19T15:55:50.302Z",
                     "newMajor": 0,
                     "newMinor": 0,
                     "updateType": "patch",
                     "isRange": true,
                     "isLockfileUpdate": true,
                     "branchName": "renovate/all-minor-patch"
                   }
                 ],
                 "packageName": "another_private_dummy",
                 "versioning": "ruby",
                 "warnings": [],
                 "sourceUrl": "https://PRIVATE/another_private_dummy",
                 "registryUrl": "https://PRIVATE",
                 "currentVersion": "0.0.1",
                 "isSingleVersion": true,
                 "fixedVersion": "0.0.1"
               }
             ],
             "lockFiles": ["Gemfile.lock"],
             "packageFile": "Gemfile"
           }
         ]
       }

...

DEBUG: branches info extended (repository=playground/renovate_bug)
       "branchesInformation": [
         {
           "branchName": "renovate/all-minor-patch",
           "prNo": 2,
           "prTitle": "feat(renovate): Update all non-major dependencies to v0.0.2",
           "result": "done",
           "upgrades": [
             {
               "datasource": "rubygems",
               "depName": "another_private_dummy",
               "displayPending": "",
               "fixedVersion": "0.0.1",
               "currentVersion": "0.0.1",
               "currentValue": "'~> 0.0.1'",
               "newValue": "'~> 0.0.1'",
               "newVersion": "0.0.2",
               "packageFile": "Gemfile",
               "updateType": "patch",
               "packageName": "another_private_dummy"
             },
             {
               "datasource": "rubygems",
               "depName": "private_dummy",
               "displayPending": "",
               "fixedVersion": "0.0.1",
               "currentVersion": "0.0.1",
               "currentValue": "'~> 0.0.1'",
               "newValue": "'~> 0.0.1'",
               "newVersion": "0.0.2",
               "packageFile": "Gemfile",
               "updateType": "patch",
               "packageName": "private_dummy"
             }
           ]
         }
       ]
```

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/bundler/extract.ts [hunk 1]
```diff
@@ -61,44 +61,69 @@ export async function extractPackageFile(
...
-            { content, packageFile, type: 'groupLine' },
-            'Bundler parsing error',
-          );
-          groupLine = 'end';
+
+    const processGroupBlock = async (
+      line: string,
+      repositoryUrl?: string,
+      trimGroupLine: boolean = false,
+    ): Promise<void> => {
```
Comment (rarkins) [inline lib/modules/manager/bundler/extract.ts:69]:
```suggestion
    async function processGroupBlock(
      line: string,
      repositoryUrl?: string,
      trimGroupLine: boolean = false,
    ): Promise<void> {
```

Comment (mustardnoise) [reply lib/modules/manager/bundler/extract.ts:69]: Unfortunatelly linter complains about this suggestion: 
```shell
 65:5  error  Move function declaration to function body root  no-inner-declarations
```

Do you think I can change the function to something like this (as suggested in the exception part of the second bullet point of [Fucntion Best Practices Doc](https://github.com/renovatebot/renovate/blob/main/docs/development/best-practices.md#functions))
```js
const processGroup = async function (
  line: string,
  repositoryUrl?: string,
  trimGroupLine: boolean = false,
): Promise<void> { ... )
```

File: lib/modules/manager/bundler/extract.spec.ts [hunk 2]
```diff
@@ -26,6 +26,9 @@ const sourceBlockWithNewLinesGemfileLock = Fixtures.get(
 const sourceBlockWithNewLinesGemfile = Fixtures.get(
   'Gemfile.sourceBlockWithNewLines',
 );
+const sourceBlockWithGroupsGemfile = Fixtures.get(
```
Comment (viceice) [inline lib/modules/manager/bundler/extract.spec.ts:29]: move to test

File: lib/modules/manager/bundler/extract.spec.ts [hunk 3]
```diff
@@ -124,4 +127,14 @@ describe('modules/manager/bundler/extract', () => {
...
     expect(res?.deps).toHaveLength(2);
   });
+
+  it('parses source blocks with groups in Gemfile', async () => {
+    fs.readLocalFile.mockResolvedValueOnce(sourceBlockWithGroupsGemfile);
+    const res = await extractPackageFile(
+      sourceBlockWithGroupsGemfile,
+      'Gemfile',
+    );
+    expect(res).toMatchSnapshot();
```
Comment (viceice) [inline lib/modules/manager/bundler/extract.spec.ts:137]: no new snapshots, read our devel guide

File: lib/modules/manager/bundler/extract.ts (hunk 1 above)
Comment (viceice) [reply lib/modules/manager/bundler/extract.ts:69]: move function to module scope, don't use nested functions (perf)

Comment (rarkins) [reply lib/modules/manager/bundler/extract.ts:69]: @viceice this function seems to manipulate variables like lineNumber so can't be easily moved

Comment (rarkins) [reply lib/modules/manager/bundler/extract.ts:69]: Moved to top of function

File: lib/modules/manager/bundler/extract.spec.ts (hunk 2 above)
Comment (rarkins) [reply lib/modules/manager/bundler/extract.spec.ts:29]: These are all outside tests already

File: lib/modules/manager/bundler/extract.ts [hunk 4]
```diff
@@ -16,12 +16,70 @@ export async function extractPackageFile(
   content: string,
   packageFile?: string,
 ): Promise<PackageFileContent | null> {
+  let lineNumber: number;
+  async function processGroupBlock(
```
Comment (viceice) [inline lib/modules/manager/bundler/extract.ts:20]: it's still nested. why not simply passing line number as argument and return the new values. we do that on other managers

Comment (rarkins) [reply lib/modules/manager/bundler/extract.ts:20]: I looked again and really feel it's not worth the refactoring because it also mutates res.deps

**Review by rarkins — APPROVED** [review]

**Pull request merged**